components/camel-solr/src/main/java/org/apache/camel/component/solr/SolrProducer.java

```diff
@@ -0,0 +1,302 @@
+            UpdateRequest updateRequest = createUpdateRequest(solrConfiguration, solrParams);
+            updateRequest.rollback();
+            updateRequest.process(solrClient, solrCollection);
+            if (solrConfiguration.isAutoCommit()) {
+                commit(exchange, solrClient, solrConfiguration, solrParams);
+            }
+        } else if (operation.equalsIgnoreCase(SolrConstants.OPERATION_OPTIMIZE)) {
+            UpdateRequest updateRequest = createUpdateRequest(solrConfiguration, solrParams);
+            updateRequest.setAction(ACTION.OPTIMIZE, true, true, 1);
+            updateRequest.process(solrClient, solrCollection);
+            if (solrConfiguration.isAutoCommit()) {
+                commit(exchange, solrClient, solrConfiguration, solrParams);
+            }
+        } else if (operation.equalsIgnoreCase(SolrConstants.OPERATION_QUERY)) {
+            query(exchange, solrClient, solrConfiguration, solrParams);
+        } else {
+            throw new IllegalArgumentException(
+                    SolrConstants.OPERATION + " header value '" + operation + "' is not supported");
+        }
+    }
+
+    private void commit(
+            Exchange exchange,
+            SolrClient solrClient,
+            SolrConfiguration solrConfiguration,
+            ModifiableSolrParams solrParams)
+            throws SolrServerException, IOException {
+        commit(exchange, solrClient, solrConfiguration, solrParams, false);
+    }
+
+    private void commit(
+            Exchange exchange,
+            SolrClient solrClient,
+            SolrConfiguration solrConfiguration,
+            ModifiableSolrParams solrParams,
+            boolean softCommit)
+            throws SolrServerException, IOException {
+        String solrCollection = exchange.getIn().getHeader(SolrConstants.COLLECTION, String.class);
+        UpdateRequest updateRequest = createUpdateRequest(solrConfiguration, solrParams);
+        updateRequest.setAction(
+                ACTION.COMMIT,
+                solrParams.getBool("waitFlush", true),
+                solrParams.getBool("waitSearcher", true),
+                softCommit);
+        updateRequest.process(solrClient, solrCollection);
+    }
+
+    private void query(
+            Exchange exchange, SolrClient solrClient, SolrConfiguration solrConfiguration, ModifiableSolrParams solrParams)
+            throws SolrServerException, IOException {
+        String solrCollection = exchange.getIn().getHeader(SolrConstants.COLLECTION, String.class);
+        SolrQuery solrQuery = new SolrQuery();
+        if (ObjectHelper.isNotEmpty(exchange.getMessage().getHeader(SolrConstants.QUERY_STRING))) {
+            solrQuery.setQuery(exchange.getMessage().getHeader(SolrConstants.QUERY_STRING, String.class));
+        } else {
+            throw new IllegalArgumentException("Query String needs to be set as header while querying Solr");
```

Comment:
`NoSuchHeaderException` here too (also in other places where we need to check for mandatory headers).

Reply:
fixed